src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignerInfoTests.netcoreapp.cs

```diff
@@ -0,0 +1,270 @@
+            Assert.Throws<CryptographicException>(() => cms.SignerInfos[0].RemoveUnsignedAttribute(fakeAttribute));
+
+            Assert.Equal(numberOfAttributes, cms.SignerInfos[0].UnsignedAttributes.Count);
+        }
+
+        [Fact]
+        public static void SignerInfo_RemoveUnsignedAttributes_RemoveWithNonMatchingData()
+        {
+            SignedCms cms = new SignedCms();
+            cms.Decode(SignedDocuments.OneRsaSignerTwoRsaCounterSigners);
+
+            int numberOfAttributes = cms.SignerInfos[0].UnsignedAttributes.Count;
+            Assert.NotEqual(0, numberOfAttributes);
+
+            AsnEncodedData fakeAttribute = new AsnEncodedData(
+                cms.SignerInfos[0].UnsignedAttributes[0].Oid,
+                cms.SignerInfos[0].UnsignedAttributes[0].Values[0].RawData.Skip(1).ToArray());
+            Assert.Throws<CryptographicException>(() => cms.SignerInfos[0].RemoveUnsignedAttribute(fakeAttribute));
+
+            Assert.Equal(numberOfAttributes, cms.SignerInfos[0].UnsignedAttributes.Count);
+        }
+
+        private static void VerifyAttributesContainsAll(CryptographicAttributeObjectCollection attributes, List<AsnEncodedData> expectedAttributes)
+        {
+            var indices = new HashSet<int>();
+            foreach (CryptographicAttributeObject attribute in attributes)
+            {
+                foreach (AsnEncodedData attributeValue in attribute.Values)
+                {
+                    int idx = FindAsnEncodedData(expectedAttributes, attributeValue);
+                    Assert.NotEqual(-1, idx);
+                    indices.Add(idx);
+                }
+            }
+
+            Assert.Equal(expectedAttributes.Count, indices.Count);
+        }
+
+        private static int FindAsnEncodedData(List<AsnEncodedData> array, AsnEncodedData data)
+        {
+            for (int i = 0; i < array.Count; i++)
+            {
+                if (AsnEncodingDataEqual(array[i], data))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private static List<AsnEncodedData> GetAllAsnEncodedDataFromAttributes(CryptographicAttributeObjectCollection attributes)
+        {
+            var ret = new List<AsnEncodedData>();
+            foreach (CryptographicAttributeObject attribute in attributes)
+            {
+                foreach (AsnEncodedData attributeValue in attribute.Values)
+                {
+                    ret.Add(attributeValue);
+                }
+            }
+
+            return ret;
+        }
+
+        private static bool AsnEncodingDataEqual(AsnEncodedData a, AsnEncodedData b)
+        {
+            return a.Oid.Value == b.Oid.Value && a.RawData.SequenceEqual(b.RawData);
+        }
+
+        private static SignedCms ReReadSignedCms(SignedCms cms)
```

Comment:
Maybe this should be `void ReReadSignedCms(ref SignedCms cms)` so no one has the expectation that it modifies the existing object?

(I expected just cms.Decode(cms.Encode()) and was surprised by the return value)

Reply:
missed this one too